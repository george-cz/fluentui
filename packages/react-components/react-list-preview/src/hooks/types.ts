import * as React from 'react';
import { List } from '../List';
import { ListItem } from '../ListItem';

export type ListSelectionState = {
  isSelected: (item: string | number) => boolean;
  toggleItem: (e: React.SyntheticEvent, id: string | number) => void;
  deselectItem: (e: React.SyntheticEvent, id: string | number) => void;
  selectItem: (e: React.SyntheticEvent, id: string | number) => void;
  clearSelection: (e: React.SyntheticEvent) => void;
  toggleAllItems: (e: React.SyntheticEvent, itemIds: string[] | number[]) => void;
  getListProps: () => Pick<React.ComponentProps<typeof List>, 'role' | 'aria-multiselectable'>;
  getListItemProps: (
    value: string | number,
  ) => Pick<React.ComponentProps<typeof ListItem>, 'tabIndex' | 'role' | 'aria-selected'>;
  selectedItems: Set<string | number>;
};

export type UseListSelectionOptions = {}; // multiselect etc
