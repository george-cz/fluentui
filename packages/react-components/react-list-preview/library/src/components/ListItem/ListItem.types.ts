import { Checkbox } from '@fluentui/react-checkbox';
import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
import { ListItemActionEvent } from '../../events/ListItemActionEvent';

export type ListItemSlots = {
  root: NonNullable<Slot<'li', 'div'>>;
  checkmark?: Slot<typeof Checkbox>;
};

export type ListItemValue = string | number;

/**
 * ListItem Props
 */
export type ListItemProps = ComponentProps<ListItemSlots> & {
  value?: ListItemValue;
  // eslint-disable-next-line @nx/workspace-consistent-callback-type -- using custom event here with no data
  onAction?: (e: ListItemActionEvent, value: ListItemValue) => void;
};

/**
 * State used in rendering ListItem
 */
export type ListItemState = ComponentState<ListItemSlots> & { selectable: boolean; navigable: boolean };
