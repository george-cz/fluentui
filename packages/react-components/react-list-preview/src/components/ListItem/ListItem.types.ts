import { Checkbox } from '@fluentui/react-checkbox';
import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type ListItemSlots = {
  root: NonNullable<Slot<'li', 'div'>>;
  checkmark?: Slot<typeof Checkbox>;
};

/**
 * ListItem Props
 */
export type ListItemProps = ComponentProps<ListItemSlots> & {
  value?: string | number;
};

/**
 * State used in rendering ListItem
 */
export type ListItemState = ComponentState<ListItemSlots> & { selectable?: boolean; hasCustomOnClick?: boolean };
