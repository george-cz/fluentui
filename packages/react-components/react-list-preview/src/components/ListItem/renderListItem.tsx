/** @jsxRuntime automatic */
/** @jsxImportSource @fluentui/react-jsx-runtime */

import { ItemLayout } from '@fluentui/react-migration-v0-v9';
import { assertSlots } from '@fluentui/react-utilities';
import type { ListItemSlots, ListItemState } from './ListItem.types';
/**
 * Render the final JSX of ListItem
 */
export const renderListItem_unstable = (state: ListItemState) => {
  assertSlots<ListItemSlots>(state);

  return (
    <state.root>
      <ItemLayout
        startMedia={state.media}
        header={state.header}
        headerMedia={state.headerMedia}
        contentWrapper={state.contentWrapper}
        contentMedia={state.contentMedia}
        endMedia={state.endMedia}
      >
        {state.root.children}
      </ItemLayout>
    </state.root>
  );
};
