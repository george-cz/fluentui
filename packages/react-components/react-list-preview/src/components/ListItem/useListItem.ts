import * as React from 'react';
import { useFocusableGroup } from '@fluentui/react-tabster';
import { getNativeElementProps, slot } from '@fluentui/react-utilities';
import type { ListItemProps, ListItemState } from './ListItem.types';
import { useListContext } from '../List/listContext';

/**
 * Create the state required to render ListItem.
 *
 * The returned state can be modified with hooks such as useListItemStyles_unstable,
 * before being passed to renderListItem_unstable.
 *
 * @param props - props from this instance of ListItem
 * @param ref - reference to root HTMLElement of ListItem
 */
export const useListItem_unstable = (props: ListItemProps, ref: React.Ref<HTMLElement>): ListItemState => {
  const { focusableItems } = useListContext();
  const focusableGroupAttrs = useFocusableGroup({ tabBehavior: 'unlimited' });

  const root = slot.always(
    getNativeElementProps('div', {
      ref,
      role: 'listitem',
      tabIndex: focusableItems ? 0 : undefined,
      ...focusableGroupAttrs,
      ...props,
    }),
    { elementType: 'div' },
  );

  const state: ListItemState = {
    components: {
      root: 'div',
    },
    root,
  };

  return state;
};
