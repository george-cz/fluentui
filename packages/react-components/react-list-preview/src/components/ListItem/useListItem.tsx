import * as React from 'react';
import { useFocusableGroup } from '@fluentui/react-tabster';
import { getIntrinsicElementProps, slot, useEventCallback, useId } from '@fluentui/react-utilities';
import type { ListItemProps, ListItemState } from './ListItem.types';
import { useListContext_unstable } from '../List/listContext';
import { Checkmark16Filled } from '@fluentui/react-icons';

const EMPTY_OBJECT = {};
const DEFAULT_ROOT_EL_TYPE = 'li';

const listPropsForSelected = {
  tabIndex: 0,
  role: 'option',
  'aria-selected': true,
  checkmark: {
    children: <Checkmark16Filled />,
  },
};

const listPropsForNotSelected = {
  tabIndex: 0,
  role: 'option',
  'aria-selected': false,
  checkmark: {
    children: null,
  },
};

function validateProperElementTypes(parentRenderedAs?: 'div' | 'ul' | 'ol', renderedAs?: 'div' | 'li') {
  if (renderedAs === 'div' && parentRenderedAs !== 'div') {
    throw new Error('ListItem cannot be rendered as a div when its parent is not a div.');
  }
  if (renderedAs === 'li' && parentRenderedAs === 'div') {
    throw new Error('ListItem cannot be rendered as a li when its parent is a div.');
  }
}

/**
 * Create the state required to render ListItem.
 *
 * The returned state can be modified with hooks such as useListItemStyles_unstable,
 * before being passed to renderListItem_unstable.
 *
 * @param props - props from this instance of ListItem
 * @param ref - reference to root HTMLLIElement | HTMLDivElementof ListItem
 */
export const useListItem_unstable = (
  props: ListItemProps,
  ref: React.Ref<HTMLLIElement | HTMLDivElement>,
): ListItemState => {
  const id = useId('listItem');
  const { value = id, onKeyDown, onClick } = props;

  const focusableItems = useListContext_unstable(ctx => ctx.focusableItems);
  const toggleItem = useListContext_unstable(ctx => ctx.selection?.toggleItem);
  const isSelectionEnabled = useListContext_unstable(ctx => !!ctx.selection);
  const isSelected = useListContext_unstable(ctx => ctx.selection?.isSelected(value));

  const parentRenderedAs = useListContext_unstable(ctx => ctx.as);
  const renderedAs = props.as || DEFAULT_ROOT_EL_TYPE;

  validateProperElementTypes(parentRenderedAs, renderedAs);

  const listItemProps = isSelected ? listPropsForSelected : listPropsForNotSelected;

  const focusableGroupAttrs = useFocusableGroup({ tabBehavior: 'limited-trap-focus' });

  const handleKeyDown: React.KeyboardEventHandler<HTMLLIElement & HTMLDivElement> = useEventCallback(e => {
    onKeyDown?.(e);

    // Compare targets to make sure this only triggers when the event is fired on the list item
    // and not on a button inside
    if (!isSelectionEnabled || e.defaultPrevented || e.target !== e.currentTarget) {
      return;
    }

    if (e.key === ' ') {
      e.preventDefault();
      toggleItem?.(e, value);
    }
  });

  const handleClick: React.MouseEventHandler<HTMLLIElement & HTMLDivElement> = useEventCallback(e => {
    onClick?.(e);

    if (!isSelectionEnabled || e.defaultPrevented) {
      return;
    }

    toggleItem?.(e, value);
  });

  const root = slot.always(
    getIntrinsicElementProps(DEFAULT_ROOT_EL_TYPE, {
      ref: ref as React.Ref<HTMLLIElement & HTMLDivElement>,
      tabIndex: focusableItems ? 0 : undefined,
      role: 'listitem',
      id: String(value),
      ...(isSelectionEnabled ? listItemProps : EMPTY_OBJECT),
      ...focusableGroupAttrs,
      ...props,
      onKeyDown: handleKeyDown,
      onClick: handleClick,
    }),
    { elementType: DEFAULT_ROOT_EL_TYPE },
  );

  const checkmark = slot.optional(props.checkmark, {
    defaultProps: { children: isSelected ? <Checkmark16Filled /> : null },
    renderByDefault: isSelectionEnabled,
    elementType: 'div',
  });

  const state: ListItemState = {
    components: {
      root: DEFAULT_ROOT_EL_TYPE,
      checkmark: 'div',
    },
    root,
    checkmark,
  };

  return state;
};
