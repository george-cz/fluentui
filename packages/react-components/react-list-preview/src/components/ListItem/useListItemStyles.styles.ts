import type { SlotClassNames } from '@fluentui/react-utilities';
import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import { createCustomFocusIndicatorStyle } from '@fluentui/react-tabster';
import type { ListItemSlots, ListItemState } from './ListItem.types';
import { tokens } from '@fluentui/react-theme';

export const listItemClassNames: SlotClassNames<ListItemSlots> = {
  root: 'fui-ListItem',
  checkbox: 'fui-ListItem__checkbox',
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    position: 'relative',
    ...createCustomFocusIndicatorStyle(
      {
        ...shorthands.outline('2px', 'solid', tokens.colorStrokeFocus2),
        ...shorthands.borderRadius(tokens.borderRadiusMedium),
      },
      { selector: 'focus' },
    ),
  },

  rootClickable: {
    display: 'flex', //TODO needed?
    cursor: 'pointer',
  },
});

/**
 * Apply styling to the ListItem slots based on the state
 */
export const useListItemStyles_unstable = (state: ListItemState): ListItemState => {
  const styles = useStyles();
  state.root.className = mergeClasses(
    listItemClassNames.root,
    styles.root,
    state.root.onClick && styles.rootClickable,
    state.root.className,
  );

  if (state.checkbox) {
    state.checkbox.className = mergeClasses(listItemClassNames.checkbox, state.checkbox?.className);
  }

  return state;
};
