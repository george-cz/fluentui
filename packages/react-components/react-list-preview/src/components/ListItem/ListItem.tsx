import * as React from 'react';
import type { ForwardRefComponent } from '@fluentui/react-utilities';
import { useCustomStyleHook_unstable } from '@fluentui/react-shared-contexts';
import { useListItem_unstable } from './useListItem';
import { renderListItem_unstable } from './renderListItem';
import { useListItemStyles_unstable } from './useListItemStyles.styles';
import type { ListItemProps } from './ListItem.types';

/**
 * ListItem component - TODO: add more docs
 */
export const ListItem: ForwardRefComponent<ListItemProps> = React.forwardRef<HTMLElement>((props, ref) => {
  const state = useListItem_unstable(props, ref);

  useListItemStyles_unstable(state);
  useCustomStyleHook_unstable('useListItemStyles_unstable')(state);
  return renderListItem_unstable(state);
});

ListItem.displayName = 'ListItem';
