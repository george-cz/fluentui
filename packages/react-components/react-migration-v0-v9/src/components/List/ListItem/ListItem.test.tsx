import * as React from 'react';
import { render } from '@testing-library/react';
import { isConformant } from '@fluentui/react-conformance';
import { ListItem } from './ListItem';
import { ListItemProps } from './ListItem.types';

describe('ListItem', () => {
  isConformant({
    Component: ListItem as React.FunctionComponent<ListItemProps>,
    displayName: 'ListItem',
    testOptions: {
      'has-static-classnames': [
        {
          props: {
            checkmark: 'test checkmark',
          },
        },
      ],
    },
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<ListItem>Default ListItem</ListItem>);
    expect(result.container).toMatchSnapshot();
  });
});
