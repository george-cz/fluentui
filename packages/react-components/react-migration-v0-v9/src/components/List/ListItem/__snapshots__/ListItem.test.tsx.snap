// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem renders a default state 1`] = `
<div>
  <li
    class="fui-ListItem"
    data-tabster="{\\"groupper\\":{\\"tabbability\\":2},\\"focusable\\":{}}"
    id="listItem8"
    role="listitem"
  >
    Default ListItem
  </li>
</div>
`;
